### PR TITLE
Fix loading dtype for plotting functions.

### LINKED ISSUE #611: Messy behavior with uint8 images
When calling `add_overlay` with an uint8 threshold, the mask is sometimes reversed. The culprit is here: https://github.com/nilearn/nilearn/blob/master/nilearn/plotting/displays.py#L532

```python
                data = np.ma.masked_inside(data, -threshold, threshold,
                                           copy=False)
```

If `threshold` is an `uint8`, then you get `threshold = 22` and `-threshold = 234` because of the underflow. There may be other cases like that here and there in the code.

Should we properly handle uint8 everywhere or simply convert the image to float systematically for plotting?

For the record, this bug happened to @KamalakerDadi when he tried to plot the Harvard Oxford atlas (since Harvard Oxford is shipped in uint8 format). So this is not an isolated case.

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Fix #611.
Supersedes #621.

I introduced a `dtype='auto'` in `check_niimg` that converts data to `int32` if it is integral and `float32` if it is real. I use it in the plotting functions so that we never end up with unsigned type. However, we are not protected from unsigned values given as threshold or vmax. I solved the current problem by casting explicitely the threshold to float but I don't know if it's the best way to do.

### CODE REVIEW

File: nilearn/_utils/niimg.py [hunk 1]
```diff
@@ -30,6 +31,26 @@ def _safe_get_data(img):
...
 
 
+def _get_data_dtype(img):
+    try:
+        return img.get_data_dtype()
+    except AttributeError:
+        return img.get_data().dtype
+
+
+def _get_target_dtype(dtype, target_dtype):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: I think that what you are looking for is somethink like dtype.kind, which is 'f' for floating point types, and 'i' for integer types.

File: nilearn/_utils/niimg.py [hunk 2]
```diff
@@ -30,6 +30,26 @@ def _safe_get_data(img):
...
+        if dtype.kind == 'i':
+            target_dtype = np.int32
+        else:
+            target_dtype = np.float32
+    if target_dtype == dtype:
+        return None
+    return target_dtype
+
+
 def load_niimg(niimg, dtype=None):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: The dtype behavior definitely needs to be documented.

File: nilearn/_utils/niimg_conversions.py [hunk 3]
```diff
@@ -176,7 +178,7 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False,
     return niimg
 
 
-def check_niimg_3d(niimg):
+def check_niimg_3d(niimg, dtype=None):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg_conversions.py]: I think that I would like the dtype behavior documented here too.

File: nilearn/_utils/niimg_conversions.py [hunk 4]
```diff
@@ -62,6 +62,11 @@ def _iter_check_niimg(niimgs, ensure_ndim=None, atleast_4d=False,
 
     target_fov: tuple of affine and shape
        If specified, images are resampled to this field of view
+
+    dtype: type
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg_conversions.py]: it's {dtype, "auto"} as a type description. "type" is something different in Python.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: Fixed.

**Comment by AlexandreAbraham** [comment]
@GaelVaroquaux I thought that this fix was included in 0.1.4. I'll create a milestone for the upcoming bugfixes.

**Comment by GaelVaroquaux** [comment]
Fuck, I forgot to merge it. Too many things happening in parallel at the sprint.

I think that I'll do another bugfix release.

**Pull request merged**